### ISSUE TITLE
Crate metadata

### ISSUE THREAD

**Issue by bjorn3** [slightknack/rust-abi-wiki#6]
> There are many benefits a standardized ABI would bring to Rust. A stable ABI enables dynamic linking between Rust crates, which would allow for Rust programs to support dynamically loaded plugins (a feature common in C/C++). Dynamic linking would result in shorter compile-times and lower disk-space use for projects, as multiple projects could link to the same dylib. For example, imagine having multiple CLIs all link to the same core library crate.

To be able to link to the same libstd using multiple rustc versions the crate metadata format would need to be stabilized. This in turn would require the stabilization of all nightly features used by libstd, as otherwise a different rustc version doesn't know what to do when it finds a reference to an unstable feature. It can't just skip it, as for example macros defined in libstd are able to use unstable features even when the macro itself is stable.

**Comment by bjorn3** [comment]
It may be nice to stabilize a small subset of the crate metadata though. This would make it possible to generate header files for many languages directly from a `.rlib`. In addition it would be possible to generate a `.rmeta` file while compiling a library written in a different language. The rest of the crate metadata could be encoded in a compiler specific way, with other compilers simply ignoring it.

**Comment by robinmoussu** [comment]
> for example macros defined in libstd are able to use unstable features even when the macro itself is stable

That's true but totally irrelevant when talking about ABI. Macros can only operate at the AST level. Once the code is compiled there are no trace of the macro anymore.

However you raise an interesting point. Currently stable means "API stable", not "ABI stable", which means that it's totally possible that a type whose layout is unstable is exposed through the ABI.

**Comment by bjorn3** [comment]
> However you raise an interesting point. Currently stable means "API stable", not "ABI stable", which means that it's totally possible that a type whose layout is unstable is exposed through the ABI.

I was considering the case where you compile a crate compiled with one rustc version against a crate compiled with another rustc version. I wasn't considering the case where you compile it against a crate compiled with the same rustc version, but later link it against a different crate version. The first case doesn't put any burden on crate developers, while the second case requires crate developers to not change certain internal details for the purpose of ABI stability.
